kvserver/rangefeed: add node level buffered sender

wenyihu6 opened the pull request:
**kvserver/rangefeed: add node level buffered sender**

This patch is the last step for reducing long-running O(ranges) goroutines in
kvserver/rangefeed. It changes the BufferedSender to use a queue to buffer
events before forwarding them to underlying grpc stream.

Closed: https://github.com/cockroachdb/cockroach/issues/129813
Release note: A new cluster setting
`kv.rangefeed.buffered_stream_sender.enabled` can now be used to allow rangefeed
to use buffered sender for all rangefeed feeds instead of buffering events
separately per client per range.

---

**kvserver/rangefeed: add capacity to node level buffered sender**

This patch adds capacity to node level buffered sender which will shut down all
registrations if the node level buffer had overflowed.

Part of: https://github.com/cockroachdb/cockroach/issues/129813
Release note: none

TODO: decide a good constant value for this, add a larger test at the kvclient side to make sure error returned here is treated as a restart signal

wenyihu6 committed:
to be squashed separated out for easier code review

wenyihu6 committed:
kvserver/rangefeed: add unbuffered registration tests

This patch adds more test cases for unbufferedRegistration.

Closed: #126560
Release note: none

wenyihu6 committed:
to be squashed separated out for easier code review

wenyihu6 committed:
kvserver/rangefeed: add newRetryErrBufferCapacityExceeded

This patch refactors the error for kvpb.RangeFeedRetryError_REASON_SLOW_CONSUMER
into newRetryErrBufferCapacityExceeded.

Part of: #126560
Release note: none

wenyihu6 committed:
kvserver/rangefeed: add rangefeed level callback

This patch adds RegisterRangefeedCleanUp to BufferedStream interface. Rangefeed
level can register a cleanup callback that will be invoked after
BufferedStream.Disconnect is called. Note that it might not be invoked
immediately during SendBufferedError.

Buffered registration has a dedicated goroutine to do cleanup. Unbuffered
registration does not have a dedicated goroutine, so it relies on BufferedSender
to invoke the callback when it disconnects.

Part of: #129814
Release note: none

wenyihu6 committed:
kvserver/rangefeed: add unbuffered registration

This patch adds unbufferedRegistration.

UnbufferedRegistration is like BufferedRangefeed but uses BufferedStream to
buffer live raft updates instead of a using buf channel and having a
dedicated per-range per-registration goroutine to volley events to underlying
grpc stream. Instead, there is only one BufferedStream for each incoming
node.MuxRangefeed rpc call. BufferedStream is responsible for buffering and
sending its updates to the underlying grpc stream in a dedicated goroutine
O(node).

Note that BufferedStreamSender is still left unimplemented. This commit doesn't
add any tests for it yet. More tests will be added in future commits of this PR.

Part of: #129814
Release note: none

wenyihu6 committed:
kvserver/rangefeed: move helper functions to registry_helpers_test

This patch moves helper functions to registry_helpers_test.

Part of: #129814
Release note: none

wenyihu6 committed:
kvserver/rangefeed: refactor registry_helpers_test to facilitate future tests

This patch refactors helper functions in registry_helpers_test.

Part of: #129814
Release note: none

wenyihu6 committed:
to be squashed separated out for easier code review

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/129937)
<!-- Reviewable:end -->

wenyihu6 committed:
to be squashed and discussed

wenyihu6 committed:
kvserver/rangefeed: add node level buffered sender

This patch is the last step for reducing long-running O(ranges) goroutines in
kvserver/rangefeed. It changes the BufferedSender to use a queue to buffer
events before forwarding them to underlying grpc stream.

Closed: #129813
Release note: A new cluster setting
`kv.rangefeed.buffered_stream_sender.enabled` can now be used to allow rangefeed
to use buffered sender for all rangefeed feeds instead of buffering events
separately per client per range.

wenyihu6 committed:
kvserver/rangefeed: add event queue

This patch adds a new data structure eventQueue which is like a queue but uses a
fixed size for the chunked linked list. Each chunk has a fixed size of 4096
elements. This implementation uses sync.Pool to reduce the number of
allocations.

pushBack, popFront, len run in constant time. removeAll runs in linear time with
respect to the number of elements in the queue. This structure is not safe for
concurrent use.

This is for future commits to include the queue in the BufferedSender to buffer
events at the node level.

Part of: #129813
Release note: none

wenyihu6 committed:
kvserver/rangefeed: add capacity to node level buffered sender

This patch adds capacity to node level buffered sender which will shut down all
registrations if the node level buffer had overflowed.

Part of: https://github.com/cockroachdb/cockroach/issues/129813
Release note: none

TODO: add a larger test at the kvclient side to make sure error returned here is treated as a restart signal

wenyihu6 committed:
to be squashed separated out for easier code reviews

wenyihu6 commented:
Closing - as we are splitting things out based on our new integration branch: https://github.com/cockroachdb/cockroach/pull/125872.
